torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,707 @@
+    nullptr,                    /* tp_call */
+    nullptr,                    /* tp_str */
+    nullptr,                    /* tp_getattro */
+    nullptr,                    /* tp_setattro */
+    nullptr,                    /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT,         /* tp_flags */
+    "Python tracer TLS",        /* tp_doc */
+    nullptr,                    /* tp_traverse */
+    nullptr,                    /* tp_clear */
+    nullptr,                    /* tp_richcompare */
+    0,                          /* tp_weaklistoffset */
+    nullptr,                    /* tp_iter */
+    nullptr,                    /* tp_iternext */
+    nullptr,                    /* tp_methods */
+    nullptr,                    /* tp_members */
+    nullptr,                    /* tp_getset */
+    nullptr,                    /* tp_base */
+    nullptr,                    /* tp_dict */
+    nullptr,                    /* tp_descr_get */
+    nullptr,                    /* tp_descr_set */
+    0,                          /* tp_dictoffset */
+    nullptr,                    /* tp_init */
+    nullptr,                    /* tp_alloc */
+    PyType_GenericNew,          /* tp_new */
+};
+
+// CPython has a more expressive set of events for tracing / profiling:
+//   https://github.com/python/cpython/blob/f291404a802d6a1bc50f817c7a26ff3ac9a199ff/Include/cpython/pystate.h#L22-L29
+// As an implementation detail they are defined as 0-7, however we don't want
+// to rely on that while bit packing. Furthermore, the CPython descriptions
+// are finer granularity than we're interested in. We do not need to
+// differentiate between a normal return and an exception (both act as a pop in
+// our replay stack), and we are not interested in `PyTrace_LINE` or
+// `PyTrace_OPCODE`. To simplify things we store our own enum when tracefunc is
+// called, and then use for all subsequent processing.
+enum TraceTag {
+    kPy_Call = 0,
+    kPy_Return,
+    kC_Call,
+    kC_Return
+};
+
+//   When we are tracing a Python program, the general procedure is to record
+// every time we enter or exit a function and later replay these events during
+// post processing. Thus, during the profiling phase we want to do the MINIMAL
+// amount of work to capture all of the information that we need; otherwise we
+// will distort the profile. (While we don't wish to be terribly inefficient
+// during post processing, we are willing to do extra fixup work in post if it
+// reduces overhead in the profiling phase.)
+//
+//   To that end, `RawEvent` (which logs calls and returns) is bitpacked to
+// reduce data stored and fit more events on a cache line. The following
+// techniques are used:
+//
+//  1) Storing `tag_` as a uint8_t rather than a TraceTag.
+//      The size of an enum, surprisingly, is not the amount of space needed
+//      to store all the fields, but rather *at least* that size.
+//      (`sizeof(TraceTag) == 2` on my system, for example.)
```

Comment:
nit: might be worth adding this assertion since we cast `uint8_t` <--> `TraceTag` all the time. but i highly doubt this invariant will break anytime soon

Reply:
I considered adding a final `kTagSize` field and static asserting, but then I'd have to handle it in the switch stmts. (Really wish C++ had a better way of asking "how many elements are in this enum?")